pageserver/src/basebackup.rs

```diff
@@ -76,17 +77,19 @@ impl<'a> Basebackup<'a> {
             (current_prev, current_last)
         };
 
+        timeline.wait_lsn(backup_lsn)?;
```

Comment:
May be move that to `if req_lsn > current_last` condition arm slightly above? Or even call that before `get_last_record_rlsn` and assert later that req_lsn <= current_last. That way we still would have a race between waiting and prev_lsn retrieval, but in cases when last_lsn didn't change during that window we would set proper prev_lsn, instead of zero.

Reply:
Hmm, can we always just use 0 as the prev LSN, if the client requested a backup at a particular LSN? The assumption here is that if the client requested a particular LSN, it's creating a compute node for read-only mode, and doesn't need the prev-LSN. Currently, we will opportunistically return a prev-LSN if we can in that case too, but if we don't need it, I think it's better to never return it.

Reply:
added a commit to demonstrate that

Reply:
We need to use backup at a particular LSN on each node restart to avoid situation where WAL is still going from safekeepers to pageserver. We can return Lsn(0) for each restart, but that would generate WAL that has wrong prev_lsn on each restart.

Other way around is not to try to set prev_lsn at pageserver but return it from `postgres --sync-safekeepers` run -- that is where we take backup_lsn in a first place.

Reply:
> We need to use backup at a particular LSN on each node restart to avoid situation where WAL is still going from safekeepers to pageserver.

Hmm, so in that case, the point of the backup LSN is "I know there is no WAL after this point, this is really the end of the timline, but if you haven't received all the WAL up to that point yet, please wait for it to arrive". I see. Ok, I'll fix this so that it works the same as before.

> Other way around is not to try to set prev_lsn at pageserver but return it from `postgres --sync-safekeepers` run -- that is where we take backup_lsn in a first place.

That is pretty appealing. Tracking the prev LSN in the page server feels pretty awkward currently.

Reply:
Ok, the logic is now the same as before. I rewrote the comment and the few lines of code involved, please take another look at the last commit in this PR.

Reply:
Yeah, looks good, thanks!

Related: https://github.com/zenithdb/zenith/issues/580